### ISSUE TITLE
Add `in-toto-sign` command line utility

### ISSUE THREAD

**Issue by lukpueh** [in-toto/in-toto#97]
in-toto `Layouts` and `Links` are both [signable](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/common.py#L72).

Let's add a command line utility to handle in-toto metadata signatures. Here are some functions I consider helpful:

- add signature, while keeping existing signature(s)
- replace existing signature(s) with new signature
- verify signature(s)

Feel free to add your own ideas.

**Comment by ShikherVerma** [comment]
Maybe it could also create a link from a link-mock file ? would that be a use case?

**Comment by ShikherVerma** [comment]
Can't assign the issue to me. Need discussion more discussion about the CLI syntax before implementing.

**Comment by lukpueh** [comment]
Good thought. I wonder though if it's really necessary. Do you have something in mind where this could be convenient?

**Comment by ShikherVerma** [comment]
Might be useful if you want to do the *link json generation and signing separately*. Like if some super paranoid guy keeps his private gpg key on a disconnected raspberry pi and only signs stuff on it ( @adarshaj i'm looking at you) or uses a [yubikey](https://www.yubico.com/why-yubico/how-yubikey-works/) and doesn't have it while generating the link json.

**Comment by lukpueh** [comment]
That definitely makes sense, but note that you can use `in-toto-run` without a key, which produces an unsigned regular `.link` file, which in turn could be signed using e.g. the here proposed command line tool.

**Comment by ShikherVerma** [comment]
Oh I see. Then this extra `link-mock to link` functionality is not required.

What would be the syntax for in-toto-sign cli.
``` bash
# Add signature
$in-toto-sign --add-sign --key bob --link building.link

# Replace signature
$in-toto-sign --replace-sign --key muck --link step.[bob-key-prefix].link

# for verification
$in-toto-sign --verify --key bob.pub --link bobthebuilder.[bob-key-prefix].link
```
For Add signature and replace signature whose key prefix goes in the link file name ? should we have key prefix if link is singed by more than 1 person?

**Comment by ShikherVerma** [comment]
What if we try to replace sign on a link which is signed by more than one person. then do remove all signatures or we add option for which signature we want to remove ?

**Comment by isachit** [comment]
Shikher are you on it?

**Comment by isachit** [comment]
like already started the programming? If yes, I can send you whatever I have done on this until now, if not let me know too.

**Comment by ShikherVerma** [comment]
@isachit you can take up this issue :+1:

**Comment by lukpueh** [comment]
Excellent questions! I had something like this in mind:

```shell
in-toto-sign --key <signing key> [--replace-all-old-signatures] [--add-keyid-infix-to-filename] <path/to/signable>
in-toto-sign --key <verifying key> --verify <path/to/signable>
```
Please think of better option/flag names.

**Comment by isachit** [comment]
```
# Add signature
$in-toto-sign --add-sign --key <path/to/bobs/key> --link <path/to/link/file>

# Replace signature
$in-toto-sign --replace-sign --key <path/to/bobs/key> --link <path/to/link/file>

# for verification
$in-toto-sign --verify --key <path/to/bobs/key> --link <path/to/link/file>
```

**Comment by isachit** [comment]
Oops sorry.

**Comment by isachit** [comment]
Alright cool! Just wanted to ask in the `<signing key>`, the path to the pub file must be specified or? @lukpueh

**Comment by lukpueh** [comment]
No you don't.

FYI, the private key actually contains private and public portion of the RSA key pair. 
For signing you usually use your private key, which let's someone who verifies your signature with your public key attribute the signature to you (because you should be the only one who has that private key).

**Comment by lukpueh** [comment]
Re *link keyid infix in file name if signed by more than one keys*:
Although currently it is technically possible to sign one link file with more than one keys, I am not sure if this is actually a use case. We provide link signing key thresholds by requiring multiple link files of for the same step, each signed by a (single) functionary.
Signing a `signable` with more than one keys is really only relevant in the case of `layouts`. 

I too briefly thought about whether it is useful to parametrize the replacement of a specific signature but I am not sure.

For now I'm satisfied with a tool that just signs a `link` or `layout` and lets me specify whether the signature should be appended or replace existing signatures and whether the keyid should be added/replace an existing keyid filename infix or not.

**Comment by isachit** [comment]
Noted.

**Comment by isachit** [comment]
The code is nearly done. Will send it asap.

**Comment by SantiagoTorres** [comment]
Fixed in #99
